cpp/oneapi/dal/algo/minkowski_distance/common.hpp

```diff
@@ -0,0 +1,148 @@
+
+#pragma once
+
+#include "oneapi/dal/detail/common.hpp"
+#include "oneapi/dal/table/common.hpp"
+
+namespace oneapi::dal::minkowski_distance {
+
+namespace task {
+namespace v1 {
+
+/// Tag-type that parameterizes entities that are used to compute distances.
+struct compute {};
+/// Alias tag-type for the compute task.
+using by_default = compute;
+} // namespace v1
+
+using v1::compute;
+using v1::by_default;
+
+} // namespace task
+
+namespace method {
+namespace v1 {
+
+struct dense {};
+/// Alias tag-type for the dense method.
+using by_default = dense;
+} // namespace v1
+
+using v1::dense;
+using v1::by_default;
+
+} // namespace method
+
+namespace detail {
+namespace v1 {
+struct descriptor_tag {};
+
+template <typename Task>
+class descriptor_impl;
+
+template <typename Float>
+constexpr bool is_valid_float_v = dal::detail::is_one_of_v<Float, float, double>;
+
+template <typename Method>
+constexpr bool is_valid_method_v = dal::detail::is_one_of_v<Method, method::dense>;
+
+template <typename Task>
+constexpr bool is_valid_task_v = dal::detail::is_one_of_v<Task, task::compute>;
+
+template <typename Task = task::by_default>
+class descriptor_base : public base {
+    static_assert(is_valid_task_v<Task>);
+
+public:
+    using tag_t = descriptor_tag;
+    using float_t = float;
+    using method_t = method::by_default;
+    using task_t = Task;
+
+    descriptor_base();
+
+    double get_degree() const;
+
+protected:
+    void set_degree_impl(double value);
+
+private:
+    dal::detail::pimpl<descriptor_impl<Task>> impl_;
+};
+
+} // namespace v1
+
+using v1::descriptor_tag;
+using v1::descriptor_impl;
+using v1::descriptor_base;
+
+using v1::is_valid_float_v;
+using v1::is_valid_method_v;
+using v1::is_valid_task_v;
+
+} // namespace detail
+
+namespace v1 {
+
+/// @tparam Float  The floating-point type that the algorithm uses for
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method Tag-type that specifies an implementation of algorithm. Can
+///                be :expr:`method::dense`.
+/// @tparam Task   Tag-type that specifies the type of the problem to solve. Can
+///                be :expr:`task::compute`.
+template <typename Float = float,
+          typename Method = method::by_default,
+          typename Task = task::by_default>
+class descriptor : public detail::descriptor_base<Task> {
+    static_assert(detail::is_valid_float_v<Float>);
+    static_assert(detail::is_valid_method_v<Method>);
+    static_assert(detail::is_valid_task_v<Task>);
+
+    using base_t = detail::descriptor_base<Task>;
+
+public:
+    using float_t = Float;
+    using method_t = Method;
+    using task_t = Task;
+
+    /// Creates a new instance of the class with the default property values.
+    descriptor() = default;
+
+    /// Creates a new instance of the class with the external property values.
+    explicit descriptor(double degree) {
```

Comment:
Why? We want set degree only with help setter's method

Reply:
It was made according to proposal:
![image](https://user-images.githubusercontent.com/54797533/119826810-1863d380-bf01-11eb-94de-a5aaa56d0e75.png)

We still can use your approach to set degree. Is there any potential problems with using this constructor?